base/iterators.jl

```diff
@@ -649,6 +649,58 @@ end
 iterate(it::Drop, state) = iterate(it.xs, state)
 isdone(it::Drop, state) = isdone(it.xs, state)
 
+
+# takewhile
+
+struct TakeWhile{I,P<:Function}
+    pred::P
+    xs::I
```

Comment:
Could you be consistent about naming of the inner iterator?

```julia
julia> lines = readlines(`sh -c "grep --line-number -A5 -E '^struct' iterators.jl | grep -E -e struct -e '::[A-Z]\$'"`)
32-element Array{String,1}:
 "77:struct Reverse{T}"
 "78-    itr::T"
 "106:struct Enumerate{I}"
 "107-    itr::I"
 "166:struct Pairs{K, V, I, A} <: AbstractDict{K, V}"
 "167-    data::A"
 "168-    itr::I"
 "263:struct Zip{Is<:Tuple}"
 "393:struct Filter{F,I}"
 "394-    flt::F"
 "395-    itr::I"
 "445:struct Rest{I,S}"
 "446-    itr::I"
 "447-    st::S"
 "504:struct Count{S<:Number}"
 "505-    start::S"
 "506-    step::S"
 "537:struct Take{I}"
 "538-    xs::I"
 "594:struct Drop{I}"
 "595-    xs::I"
 "653:struct Cycle{I}"
 "654-    xs::I"
 "691:struct Repeated{O}"
 "692-    x::O"
 "726:struct ProductIterator{T<:Tuple}"
 "727-    iterators::T"
 "848:struct Flatten{I}"
 "849-    it::I"
 "932:struct PartitionIterator{T}"
 "933-    c::T"
 "957:struct IterationCutShort; end"

julia> count(x -> occursin("xs::", x), lines)
3

julia> count(x -> occursin("itr::", x), lines)
5
```

`xs` is popular but using `itr` would be most consistent with everything else.

(I know this is an implementation detail but it helps me implementing https://tkf.github.io/Transducers.jl/dev/manual/#Transducers.eduction)

Reply:
The problem is that there is a lot of inconsistency between iterator as `iterable` and `iterator` as `state` (iteration state) eg arg1 and arg2 of `iterate/2`. In `C#` there is clearly `IEnumerable` and `IEnumerator` interface. I am not sure `Linq` would be so great otherwise.

Most people are only data poweruser and are not so aware. Julia should inforce this too i think. Following this advise will lead to augment confusion, i don't think it's a good way.

I wrote my code with either `ibl`/`itr` either `xs`/`itr` when `ibl` is indexable. It was relaxed for Base. But it works ... everywhere, so far :smiley:

Reply:
> The problem is that there is a lot of inconsistency between iterator as `iterable` and `iterator` as `state` (iteration state)

I know the difference. But please observe that all the `struct`s above with `xs` and `itr` are the same thing; i.e., they are fed to the first argument of `iterate`. Consider `itr` an abbreviation of iterable if that's your preference.

> In `C#` there is clearly `IEnumerable` and `IEnumerator` interface.

I'm not familiar with C# but this seems to be irrelevant in Julia because the iterator interface in Julia is formalized in terms of methods without side-effect; i.e., there is no mutable "iterator" object. (My understanding is that IEnumerable and IEnumerator in C# are iterable and iterator in Python)

Reply:
I don't understand your point. You ask me to be consistent on a change over 5% of iterators constating base isn't about over 95%. Names used are the same as take/while, the nearer in semantics. So yes this is consistent.

It could be a good thinks to homogenize first. But that's not done and that's not the purpose of adding takewhile, dropwhile. May be submit a pr for that. May be better, participate to an iterator julep https://github.com/JuliaLang/julia/issues/33239

Same pattern in python, java, c# it seems for distinction between `iterable`/`iterator` eg. `enumerable`/`enumerator`.

By the way, sorry to say that bikeshedding for hour is not the way i will develop open source software. Too much work, too few time to enjoy life

Reply:
It's not bikeshedding. It affects code dealing with iterators, as I mentioned in the first comment.

Reply:
The field names are already not consistent, so I'd suggest just changing them all in a separate PR after this is merged.

I don't understand how it's useful to access the iterator "inside" another. Does that operation mean anything in terms of the iteration abstraction?

Reply:
Yes, renaming them and/or adding a proper API (e.g., `parent`) makes sense.

> Does that operation mean anything in terms of the iteration abstraction?

It's certainly unnecessary for iteration abstraction.  I'm using it for converting "iterator transforms" like `Base.Generator` to transducers which unfortunately requires to get into the concrete implementation details:

https://github.com/tkf/Transducers.jl/blob/3cd6463ea08cd475b2020a6e8deca244a105d20e/src/comprehensions.jl#L89-L102

FYI my another comment on iterator-transducer conversion: https://github.com/JuliaLang/julia/pull/33027#issuecomment-524529959

Reply:
Yes, renaming them and/or adding a proper API (e.g., `parent`) makes sense.

> Does that operation mean anything in terms of the iteration abstraction?

It's certainly unnecessary for iteration abstraction.  I'm using it for converting "iterator transforms" like `Base.Generator` to transducers which unfortunately requires to get into the concrete implementation details:

https://github.com/tkf/Transducers.jl/blob/3cd6463ea08cd475b2020a6e8deca244a105d20e/src/comprehensions.jl#L89-L102

FYI my another comment on iterator-transducer conversion: https://github.com/JuliaLang/julia/pull/33027#issuecomment-524529959